src/accumulator/proof.rs

```diff
@@ -241,6 +243,291 @@ impl Proof {
+        // Combine the hashes with the targets.
+        let orig_targets_with_hash: Vec<(u64, sha256::Hash)> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(cached_del_hashes.into_iter())
+            .collect();
+
+        // Attach positions to the proof.
+        let proof_pos = get_proof_positions(
+            &self.targets,
+            before_num_leaves,
+            util::tree_rows(before_num_leaves),
+        );
+        let proof_with_pos = proof_pos
+            .clone()
+            .into_iter()
+            .zip(self.hashes.clone())
+            .collect();
+
+        // Remap the positions if we moved up a after the addition row.
+        let targets_after_remap = Proof::maybe_remap(
+            before_num_leaves,
+            adds.len() as u64,
+            orig_targets_with_hash.clone(),
+        );
+        let mut final_targets = targets_after_remap.clone();
+        let mut new_nodes_iter = new_nodes.iter();
+        let proof_with_pos =
+            Proof::maybe_remap(before_num_leaves, adds.len() as u64, proof_with_pos);
+        // remembers is an index telling what newly created UTXO should be cached
+        for remember in remembers {
+            let remember_target: Option<&sha256::Hash> = adds.get(remember as usize);
+            if let Some(remember_target) = remember_target {
+                let node = new_nodes_iter.find(|(_, hash)| *hash == *remember_target);
+                if let Some((pos, hash)) = node {
+                    final_targets.push((*pos, *hash));
+                }
+            }
+        }
+
+        final_targets.sort();
+        // Move up positions that need to be moved up due to the empty roots
+        // being written over.
+        let num_leaves = before_num_leaves + (adds.len() as u64);
+        let (mut new_target_pos, target_hashes): (Vec<_>, Vec<_>) =
+            final_targets.clone().into_iter().unzip();
+        let mut new_proof_pos = proof_pos;
+
+        for node in to_destroy {
+            (new_target_pos, _) =
+                Proof::calc_next_positions(&vec![node], &final_targets, num_leaves, true)?;
+            (new_proof_pos, _) =
+                Proof::calc_next_positions(&vec![node], &proof_with_pos, num_leaves, true)?;
+        }
+        // Add in proof_hashes to the new_nodes as some needed hashes
+        // will be in the proof hashes.
+        let leaves_after_add = ((before_num_leaves as usize) + adds.len()) as u64;
+        // Grab all the new nodes after this add.
+        let needed_proof_positions = util::get_proof_positions(
+            &new_target_pos,
+            leaves_after_add,
+            util::tree_rows(leaves_after_add),
+        );
+        let mut new_proof: Vec<_> = new_proof_pos.into_iter().zip(self.hashes).collect();
+        for pos in needed_proof_positions {
+            for (new_node_pos, hash) in new_nodes.iter() {
+                if *new_node_pos == pos {
```

Comment:
There should be an error check here for when we don't find the needed position.

The vec of nodes should always contain all the needed positions

Reply:
When I add an else clause with an error, it breaks. I am not sure why.

Reply:
Could you post the code you used and the error that occurs when you made the change?

Reply:
Sure. This is the `for` loop in line 350.  
```rust
        for pos in needed_proof_positions {
            for (new_node_pos, hash) in new_nodes.iter() {
                if *new_node_pos == pos {
                    new_proof.push((pos, *hash));
                } else {
                    return Err(format!("Missing position {pos}"));
                }
            }
        }
```

Reply:
Ah ok so this is related to https://github.com/mit-dci/rustreexo/pull/17/files#r1029246624

For positions: `[0, 1, 7, 10]`, you need `[6, 11, 17, 18, 20, 27]`. In that loop, it'll try to grab every single one of the newly created positions starting from `6` to `27`.

However, `6` isn't part of the newly created positions. It's part of the old proofs that existed before this block.

In the go version, I'm creating a whole new proof so what I did was I added the old proofs to the `new_nodes`. In yours, the `new_proof` already contains `6`, along with `17` and `18`.

So I guess you need to filter out what you already have from the `needed_proof_positions`.

`needed_proof_positions` = `[6, 11, 17, 18, 20, 27]` but since you already have `[6, 17, 18]` in `new_proof`, `needed_proof_positions` can just be `[11, 20, 27]`.